src/main/java/hudson/remoting/AgentFailuresAnalyzer.java

```diff
@@ -0,0 +1,220 @@
+package hudson.remoting;
+
+import org.jvnet.animal_sniffer.IgnoreJRERequirement;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.lang.management.LockInfo;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MonitorInfo;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+/**
+ * Generates diagnosis information (threaddump + stacktrace) when the pingThread fails
+ *
+ * <p>
+ * Useful to troubleshoot agent disconnection failures
+ *
+ * @since 2.55
+ */
+public class AgentFailuresAnalyzer {
+
+    /**
+     * Channel we are going to study
+     */
+    private Channel c;
+
+    /**
+     * The cause of the ping failure
+     */
+    private Throwable cause;
```

Comment:
Possibly make those fields final if the goal is to make that analyzing class immutable